Team data --> Master

harishbommakanti opened the pull request:
Review of the team data page, where the user can view/enter pit/match scouting answers.

harishbommakanti committed:
Initial team data commit

- added a js page for teamData
- modal box added, shows both or which one (event or season) is unselected and which pages the user can go to
- otherwise, an instruction text appears at the top of the page and displays what the page does
- next steps are to display teams for a given event for a specific season to move on with team data

harishbommakanti committed:
team radio list added

- teams in the current event and season are displayed in the radio list
- now, need to work on adding a cookie for storing the current team name and layout of the page, such as the tabbed sections and the w3css grid

harishbommakanti committed:
added onchange method for team radio list

- added helper methods for getting curr team and curr numbers
- onchange method sets a cookie for curr team and sets the title for the page

harishbommakanti committed:
Tabbed design for the team data added

- added a w3css page from w3schools  to help with formatting the tabs
- added html elements with div layout recommended on w3schools, added js code to switch tabs/tab formatting
- now need to work on actually storing/retrieving pit scouting questions/answers, etc

harishbommakanti committed:
Added table with input boxes for pit scouting

- added html/js for pit scouting
- need to work on reading/storing the questions based on team via cookies and showing them again in the table

harishbommakanti committed:
Added tag to more clearly see curr team

- added html tag and code to display the curr team above the table more clearly

harishbommakanti committed:
localStorage code to store pit answes

- used window.localStorage() to keep track of pit answers and questions

harishbommakanti committed:
Altered localStorage format

- changed localStorage format so the weird table ID's that I had setup before. Each team has only 1 obj per event per season
- Altered text on page
- Previously saved questions will show up in the input boxes

harishbommakanti committed:
attempts at styling pit scouting table

- some attempts at styling the pit scouting table, but will worry about functionality first

harishbommakanti committed:
Minor debugging, edge cases

- When user changed questions and went back, form would fill up with 'undefined,' should be fixed now
- some minor things like js.clear on load and stuff, moved a method to commonmethodhelpers
- deleted loadPit( ), all that was needed was displayTableQuestions()
- next steps are more debugging and moving on to match data

harishbommakanti committed:
Added match data table

- added match data table that loads elements from season config cookies into a table with similar format to pit question table
- need to work on adding a new radio list to show match data by numbers

harishbommakanti committed:
Added most of code for match number radio list

- User can enter match number for current scouting report in the match scouting tab
- need to iron out some of the finer details like pressing the enter key and style

harishbommakanti committed:
prev stored match nums will load

- added method to load previously stored match nums into a radio list for a team
- need to work on storing that curr checked team num in a local storage object

harishbommakanti committed:
Curr checked match saved in local storage

- {'curr checked match' : '/season/event/team/match number'}
- added another method to parse that object and return the match number given a season, event, and team

harishbommakanti committed:
Added method to save match answers

- user can submit answers to element table
- need to wory about loading saved answers now during an onchange in matches

harishbommakanti committed:
Match table loads previous data

- Table loads previous data based on match number
- Need to work on other features like showing data from other teams in the same match (maybe?)

harishbommakanti committed:
Minor changes

- rearranging methods, fixing an edge case, removing some html paragraph text

harishbommakanti committed:
Added cumulative data table for pit scouting

- All data is displayed in the pit scouting tab to more easily compare teams against others

harishbommakanti committed:
Added button to delete checked match num

- title, minor code cleanup

harishbommakanti committed:
Added allMatchData table

- added table to see data from previously saved teams for a selected match

mshafer1 commented on scripts/commonMethodHelpers.js:
```diff
@@ -15,4 +15,53 @@ function getCookie(name)
+{
+    var cookieEv = getCookie('currCheckedEvent');
+    return cookieEv.substring(cookieEv.lastIndexOf('/')+1);
+}
```
good use of the string template and the substring method :)

mshafer1 commented on scripts/commonMethodHelpers.js:
```diff
@@ -15,4 +15,53 @@ function getCookie(name)
+    var cvalue = getCookie(cname);
+    if (cvalue.trim().length == 0) return '';
+
+    var result = JSON.parse(cvalue);
```
you're returning either a string or an object.

JS allows this, but it makes it difficult while your trying to use this function.

You should always return an object. You can choose whether to return an empty object
`return  {}`
or null
`return Null`

likely, the empty object makes more sense, as when the cookie does not exist, there are not teams (i.e., empty object -> empty list of teams).

mshafer1 commented on scripts/commonMethodHelpers.js:
```diff
@@ -15,4 +15,53 @@ function getCookie(name)
+function getCurrTeamNumber()
+{
+    var cvalue = getCookie('currCheckedTeam'); //contains number of team
+    var season = getCookie('currCheckedSeason');
```
you have these all over the place.

What if you ever decide to change the cookie name?

This is just a slight improvement, but it would make going back and writing automated tests easier. You might want to consider putting "GetCurrentSeason" as a method in the helpers, and update the many callers to it (same for season and teamNumber).

mshafer1 replied:
you added the methods, but didn't use them . . .

mshafer1 commented on scripts/commonMethodHelpers.js:
```diff
@@ -15,4 +15,53 @@ function getCookie(name)
+    var season = getCookie('currCheckedSeason');
+    var event = getCurrEvent();
+
+    if (cvalue.trim().length == 0 || season.length==0 || event.length==0) return '';
```
wouldn't 0 make more sense as a default?

mshafer1 commented on scripts/commonMethodHelpers.js:
this will return '' for an empty string, just worth noting if that is a case you expected to be handled.

mshafer1 commented on styles.css:
```diff
@@ -178,4 +179,27 @@ table, th, td
+.q-td
+{
+
+}
```
empty rules are invalid. You don't need to define these if you're not putting anything in them here.

mshafer1 commented on team-data.html:
```diff
@@ -1,9 +1,11 @@
 <!DOCTYPE html>
 <meta charset="UTF-8">
 <link rel="stylesheet" type="text/css" href="styles.css"> <!--css sheet-->
+<link rel="stylesheet" type="text/css" href="w3css.css"> <!--css sheet-->
```
:)

mshafer1 commented on team-data.html:
```diff
@@ -26,6 +28,101 @@ <h1 id = "title">FRC 6321 Scouting</h1>
+                    <div class=" w3-half w3-center w3-bottombar-maroon w3-hover-khaki w3-padding tabs">Pit Scouting</div>
+                </a>
+                <a href="javascript:void(0)" onclick='displayTabContent(event, "matchTabContent");'>
+                    <div class=" w3-half w3-center w3-bottombar-maroon w3-hover-khaki w3-padding tabs">Match Scouting</div>
```
really liked how you swapped the bottombar - it looks really good and helps me know which tab I'm on.

mshafer1 commented on team-data.html:
```diff
@@ -26,6 +28,101 @@ <h1 id = "title">FRC 6321 Scouting</h1>
+            <div id='pitTabContent' style='display:none' class='w3-container w3-animate-fading tabContent'>
+                <h3>Pit Scouting</h3>
+                <h4>Here, enter and view answers to pit scouting questions. If you need to edit the questions, go to the 
+                    <a href="season-config.html">Season Configuration</a> page.<br><br>
```
good :) provide a local path (right here) to do necessary actions.

mshafer1 commented on scripts/teamDataScript.js:
```diff
@@ -0,0 +1,617 @@
+    if (season.trim().length == 0 || event.trim().length == 0) return;
+
+    var teamsJSON = getTeams(season,event);
+    var keys = Object.keys(teamsJSON);
```
yeah - you're expecting an object here. so `getTeams` should always return an object (see comment there).

mshafer1 commented on scripts/teamDataScript.js:
```diff
@@ -0,0 +1,617 @@
+
+    loadPit();
+    loadMatch();
+    localStorage.setItem('currCheckedMatch','');
```
make storing the match a method. (see comment about loading cokies and testing above).

mshafer1 commented on scripts/teamDataScript.js:
```diff
@@ -0,0 +1,617 @@
+      tablinks[i].className = tablinks[i].className.replace(" w3-bottombar-gold", " w3-bottombar-maroon");
+    }
+    document.getElementById(tabName).style.display = "block";
+    evt.currentTarget.firstElementChild.className += " w3-bottombar-gold";
```
little bit of jQuery can shorten this up:

```js
   x = document.getElementsByClassName("tabContent");
    for (i = 0; i < x.length; i++) {
      x[i].style.display = "none";
    }
```
becomes:
```js
$('.tabContent').hide();
```

and 

```js
    tablinks = document.getElementsByClassName("tabs");
    for (i = 0; i < x.length; i++) {
      tablinks[i].className = tablinks[i].className.replace(" w3-bottombar-gold", " w3-bottombar-maroon");
    }
```

becomes
```js
$('.tabs').removeClass('w3-bottombar-gold').addClass('w3-bottombar-marron');
```

and lastly,
```js
document.getElementById(tabName).style.display = "block";
```

becomes
```js
$('#' + tabName).show();
```

(just a thought :smile:  )

mshafer1 commented on scripts/teamDataScript.js:
```diff
@@ -0,0 +1,617 @@
+    var season = getCookie('currCheckedSeason');
+    var cname = season + ' pitQuestions';
+    var event = getCurrEvent();
+    var questions = getCookie(cname).split('Ω'); //Ω is separator for questions
```
good that you recognize you need to clarify the 'Ω' with a comment.
Event better would be to store the character in a constant in a shared file (like cookie helpers) like:
```js
const COOKIE_STRING_SEPERATOR = 'Ω';
```
then when you re-use it here, you can refer to it by name and the comment becomes unneeded because the code itself shows why you're splitting on that.

mshafer1 commented on scripts/teamDataScript.js:
```diff
@@ -0,0 +1,617 @@
+
+    var all_team_QA_pairs_obj = [];
+    $('#pitQuestionTable tr').each(function(){ //for each row
+        var question = $(this).find('td:first').text();
```
I wonder if we should put 'question' and 'answer' as classes on the row, so this would become:
`$(this).find('.question').text()`
and
`$(this).find('.answer').val();`

mshafer1 commented on scripts/teamDataScript.js:
```diff
@@ -0,0 +1,617 @@
+
+        var pair = {};
+        pair[question] = answer;
+        if (JSON.stringify(pair).localeCompare('{}') != 0) //first row reads {} as its reading the table header
```
`JSON.stringify(pair) != '{}'`

localCompare is for sorting (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/localeCompare)

mshafer1 commented on scripts/teamDataScript.js:
```diff
@@ -0,0 +1,617 @@
+    var objName = pitAnswerObjName(season,event,team);
+    localStorage.setItem(objName,JSON.stringify(all_team_QA_pairs_obj));
+    
+    $('#pitTableConfirmation').text('Answers saved.');
```
set this to "saving..." at the start so that it shows something besides 'saved' if an error occurs.

mshafer1 commented on scripts/teamDataScript.js:
```diff
@@ -0,0 +1,617 @@
+    $('#allTeamsPitAnswers').html('');
+    $('#allPitAnswersTitle').text(`All Pit Scouting Data for the '${event} event'.`);
+
+    var tableHeader = '<tr><th></th>';
```
what does the emtpy `<th></th>` do?

harishbommakanti replied:
It leaves the top-left most cell blank as the first column of the table is just for team numbers and not for pit questions.

mshafer1 replied:
go ahead and label that `Team #` - if we ever add filtering tooling (like sorting or showing only selected teams), a labeled header would be good.

mshafer1 commented on scripts/teamDataScript.js:
```diff
@@ -0,0 +1,617 @@
+    $('#allPitAnswersTitle').text(`All Pit Scouting Data for the '${event} event'.`);
+
+    var tableHeader = '<tr><th></th>';
+    for (var i = 0; i < questionsArr.length-1; i++)
```
why `-1`?
do you not want the last question?

harishbommakanti replied:
QuestionsArr is set up so that each question had the format 'question + Ω ', so when splitting, the last item would always be ''. Letting the loop go to qArr.length-1 and not qArr.length-2 meant adding a blank column.

mshafer1 replied:
hmmm... well, this is functionally correct.

But think for a moment, does `questionsArr` tell you that it contains a list of questions + a blank string?

Or for looping through `questionsArr.length - 1` that you're going to get all the questions?

A better solution would be to cause the questionsArr to actually be a list of all questions (plus nothing extra).
I put a comment on when you get the QuestionsArr on how to do this (just trim off the extra separator from the end first).

mshafer1 commented on scripts/teamDataScript.js:
```diff
@@ -0,0 +1,617 @@
+        var element = currCName.substring(currCName.lastIndexOf('/')+1);
+        var value = getCookie(currCName);
+
+        var inputHTML = `<input type="text" class="q-input">`;
```
we should setup where things (which will not be all) that we want to track that are numbers are `type="number"` to allow for up/down buttons or the numeric input.

mshafer1 commented on scripts/teamDataScript.js:
```diff
@@ -0,0 +1,617 @@
+
+        var inputHTML = `<input type="text" class="q-input">`;
+        var elementText = `${element}`;
+        var newTRHtml = '<tr><td>' + elementText + '</td><td>' + inputHTML + "</td></tr>";
```
use a class or something to increassing the `:after` spacing (i.e., add some padding to the bottom of these for easier target range when trying to select one).

mshafer1 commented on scripts/teamDataScript.js:
```diff
@@ -0,0 +1,617 @@
+        var elementText = `${element}`;
+        var newTRHtml = '<tr><td>' + elementText + '</td><td>' + inputHTML + "</td></tr>";
+        
+        if ($('#matchQuestionTable').html().indexOf(elementText) < 0) //if table doesn't already have it
```
you shouldn't have to check this, scoring element names are unique, right?

mshafer1 commented on scripts/teamDataScript.js:
```diff
@@ -0,0 +1,617 @@
+            if (parseInt(currNum) == parseInt(newMatchNum))
+            {
+                alreadyStored = true;
+                break;
```
breaking out of loops is considered bad. Sometimes you can't avoid it, but in this case, change the loop condition to `i < numsArr.length && !alreadyStored` and remove this break so that when you first come to the loop you "see" that it is looping through the array till `alreadyStored` is determined, and the contents of the loop are only focused on how to determine that (and not on breaking out).

mshafer1 commented on scripts/teamDataScript.js:
```diff
@@ -0,0 +1,617 @@
+
+    var element_answer_pairs = [];
+    $('#matchQuestionTable tr').each(function(){ //for each row
+        var element = $(this).find('td:first').text();
```
see other comment about classes for answer and question.

mshafer1 commented on scripts/teamDataScript.js:
```diff
@@ -0,0 +1,617 @@
+function getCurrMatchNum(season,event,team) //returns number of match
+{
+    var value = localStorage.getItem('currCheckedMatch');
+    if (value.trim().length == 0)
```
be consistent with styles. Every other `if` statement has brackets.

mshafer1 reviewed (changes requested):
Overall behavior is pretty good. Couple of thoughts,

If you make match number an input, then as I'm sitting there trying to scout a bunch of matches, I can just tweak the number and select the team I'm going to scout. - i.e., I think having to add to a list and then select is going to make it require a lot of front work compared to just a number input. You could do the list. If you do, I would suggest not making it team specific, but event specific (and maybe make an easy way to add them all (like, "total number of matches" button).

harishbommakanti committed:
PR fix: common method helpers

harishbommakanti committed:
PR fix: +setCurrMatch(), refined savePit()

mshafer1 commented on scripts/teamDataScript.js:
```diff
@@ -0,0 +1,617 @@
+    var season = getCookie('currCheckedSeason');
+    var event = getCurrEvent();
+    var cname = season + ' pitQuestions';
+    var questionsArr = getCookie(cname).split('Ω');
```
`getCookie(cname).trim('Ω').split('Ω')` will yield an array without the trailing blank question is is not really there.

harishbommakanti replied:
That wasn't working so I just used getCookie(cname).split('Ω').slice(0,-1). After looking it up it looks like trim() doesn't accept any parameters.

mshafer1 replied:
oh, I mixed up my languages again . . . (https://www.w3schools.com/python/ref_string_strip.asp),  `slice` is a good other version :)

harishbommakanti committed:
PR fix: table styling

harishbommakanti committed:
PR fix: -allMatchStuff,

- commented out most of the match stuff
- math input is now an input alongside other element inputs
- all event match data table shows all teams/matches events
- now need to work on sorting by team and match

harishbommakanti committed:
PR fix: match table is sorted by team

harishbommakanti committed:
PR fix: minor code fixes

mshafer1 commented:
Noticed couple of more things:

This time, when i first open the team-data page, I something had broken because I didn't see anything scouting fields. You should default one of the tabs visible/highlighted.

Good spacing around the input.

harishbommakanti commented:
Do you mean that either the Pit/Match Scouting tab should be open by default when the window loads?

mshafer1 commented:
Yes, probably the match

harishbommakanti committed:
PR fix: matchTab opens by default

mshafer1 commented on scripts/commonMethodHelpers.js:
don't comment out code, remove it and make a commit :)

mshafer1 commented on scripts/teamDataScript.js:
```diff
@@ -0,0 +1,435 @@
+{
+    var i, x, tablinks;
+    $('.tabContent').hide();
+    $('.tabs').removeClass('w3-bottombar-gold').addClass('w3-bottombar-marron');
```
`marron`? I think that's my typo . . .

mshafer1 commented on scripts/teamDataScript.js:
```diff
@@ -0,0 +1,435 @@
+//basically copy pasted from w3schools, onclick for a change in tabs
+function displayTabContent(evt, tabName)
+{
+    var i, x, tablinks;
```
I don't think you're using any of these variables anymore. (iff I'm right, remove them).

harishbommakanti committed:
PR fix: used methods, made more constants

- used getCurrSeason() more throughout all scripts
- made constants for the greek letters
- commented out getCurrMatch() method

harishbommakanti committed:
PR fix: edge case where fields are blank

- fixed case where pit or match data fields are blank
